classes/client/azure_client.php

```diff
@@ -0,0 +1,282 @@
+<?php
+/**
+ * Azure object storage client.
+ *
+ * @package    mahara
+ * @subpackage module.objectfs
+ * @author     Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace module_objectfs\client;
+
+defined('INTERNAL') || die();
+
+require_once(get_config('docroot') . 'module/azure/vendor/autoload.php');
```

Comment:
Look at what we do in moodle: https://github.com/catalyst/moodle-tool_objectfs/blob/master/classes/client/azure_client.php#L32

This allows us to gracefully handle not having the required modules and display warnings. On mahara without module/aws i get the following fatal error:

`
[Mon Jan 22 05:13:39.767464 2018] [:error] [pid 22004] [client 192.168.124.191:38896] [WAR] 2f (module/objectfs/classes/client/s3_client.php:15) require_once(/var/www/siteroot/htdocs/module/aws/sdk/aws-autoloader.php): failed to open stream: No such file or directory
[Mon Jan 22 05:13:39.767518 2018] [:error] [pid 22004] [client 192.168.124.191:38896] Call stack (most recent first):
[Mon Jan 22 05:13:39.767525 2018] [:error] [pid 22004] [client 192.168.124.191:38896]   * log_message("require_once(/var/www/siteroot/htdocs/module/aws/s...", 8, true, true, "/var/www/siteroot/htdocs/module/objectfs/classes/c...", 15) at /var/www/siteroot/htdocs/lib/errors.php:521
[Mon Jan 22 05:13:39.767531 2018] [:error] [pid 22004] [client 192.168.124.191:38896]   * error(2, "require_once(/var/www/siteroot/htdocs/module/aws/s...", "/var/www/siteroot/htdocs/module/objectfs/classes/c...", 15, array(size 9)) at /var/www/siteroot/htdocs/module/objectfs/classes/client/s3_client.php:15
[Mon Jan 22 05:13:39.767547 2018] [:error] [pid 22004] [client 192.168.124.191:38896]   * require_once() at /var/www/siteroot/htdocs/module/objectfs/classes/client/s3_client.php:15
[Mon Jan 22 05:13:39.767552 2018] [:error] [pid 22004] [client 192.168.124.191:38896]   * require_once("/var/www/siteroot/htdocs/module/objectfs/classes/c...") at /var/www/siteroot/htdocs/module/objectfs/lib.php:13
[Mon Jan 22 05:13:39.767557 2018] [:error] [pid 22004] [client 192.168.124.191:38896]   * require_once("/var/www/siteroot/htdocs/module/objectfs/lib.php") at /var/www/siteroot/htdocs/lib/mahara.php:1601
[Mon Jan 22 05:13:39.767563 2018] [:error] [pid 22004] [client 192.168.124.191:38896]   * safe_require("module", "objectfs") at /var/www/siteroot/htdocs/lib/upgrade.php:1073
[Mon Jan 22 05:13:39.767567 2018] [:error] [pid 22004] [client 192.168.124.191:38896]   * validate_plugin("module", "objectfs") at /var/www/siteroot/htdocs/admin/extensions/plugins.php:106
[Mon Jan 22 05:13:39.767572 2018] [:error] [pid 22004] [client 192.168.124.191:38896] 
[Mon Jan 22 05:13:39.767616 2018] [:error] [pid 22004] [client 192.168.124.191:38896] PHP Fatal error:  require_once(): Failed opening required '/var/www/siteroot/htdocs/module/aws/sdk/aws-autoloader.php' (include_path='/var/www/siteroot/htdocs/webservice/libs/zend:/var/www/siteroot/htdocs/webservice/libs/zend:/var/www/siteroot/htdocs/lib/:/var/www/siteroot/htdocs/lib/pear/:.:/usr/share/php:/usr/share/pear') in /var/www/siteroot/htdocs/module/objectfs/classes/client/s3_client.php on line 15
[Mon Jan 22 05:13:39.767633 2018] [:error] [pid 22004] [client 192.168.124.191:38896] PHP Stack trace:
[Mon Jan 22 05:13:39.767647 2018] [:error] [pid 22004] [client 192.168.124.191:38896] PHP   1. {main}() /var/www/siteroot/htdocs/admin/extensions/plugins.php:0
[Mon Jan 22 05:13:39.767656 2018] [:error] [pid 22004] [client 192.168.124.191:38896] PHP   2. validate_plugin() /var/www/siteroot/htdocs/admin/extensions/plugins.php:106
[Mon Jan 22 05:13:39.767745 2018] [:error] [pid 22004] [client 192.168.124.191:38896] PHP   3. safe_require() /var/www/siteroot/htdocs/lib/upgrade.php:1073
[Mon Jan 22 05:13:39.767759 2018] [:error] [pid 22004] [client 192.168.124.191:38896] PHP   4. require_once() /var/www/siteroot/htdocs/lib/mahara.php:1601
[Mon Jan 22 05:13:39.767767 2018] [:error] [pid 22004] [client 192.168.124.191:38896] PHP   5. require_once() /var/www/siteroot/htdocs/module/objectfs/lib.php:13

`

Reply:
This should also be done for both the AWS client and the azure client.

Reply:
Just pulling in this branch, causes the plugin to fatal error since its requiring the azure module but it should work without it

> [Wed Jan 24 15:06:10.605781 2018] [:error] [pid 11047] [client 127.0.0.1:45194] [WAR] d4 (module/objectfs/classes/client/azure_client.php:15) require_once(/home/marcus/workspace/cqu/cqu-mahara/htdocs/module/azure/vendor/autoload.php): failed to open stream: No such file or directory, referer: http://cqu-mahara-dev.marcus.local/admin/extensions/plugins.php
[Wed Jan 24 15:06:10.606081 2018] [:error] [pid 11047] [client 127.0.0.1:45194] Call stack (most recent first):, referer: http://cqu-mahara-dev.marcus.local/admin/extensions/plugins.php
[Wed Jan 24 15:06:10.606088 2018] [:error] [pid 11047] [client 127.0.0.1:45194]   * log_message("require_once(/home/marcus/workspace/cqu/cqu-mahara...", 8, true, true, "/home/marcus/workspace/cqu/cqu-mahara/htdocs/modul...", 15) at /home/marcus/workspace/cqu/cqu-mahara/htdocs/lib/errors.php:521, referer: http://cqu-mahara-dev.marcus.local/admin/extensions/plugins.php
[Wed Jan 24 15:06:10.606092 2018] [:error] [pid 11047] [client 127.0.0.1:45194]   * error(2, "require_once(/home/marcus/workspace/cqu/cqu-mahara...", "/home/marcus/workspace/cqu/cqu-mahara/htdocs/modul...", 15, array(size 8)) at /home/marcus/workspace/cqu/cqu-mahara/htdocs/module/objectfs/classes/client/azure_client.php:15, referer: http://cqu-mahara-dev.marcus.local/admin/extensions/plugins.php
[Wed Jan 24 15:06:10.606096 2018] [:error] [pid 11047] [client 127.0.0.1:45194]   * require_once() at /home/marcus/workspace/cqu/cqu-mahara/htdocs/module/objectfs/classes/client/azure_client.php:15, referer: http://cqu-mahara-dev.marcus.local/admin/extensions/plugins.php
[Wed Jan 24 15:06:10.606099 2018] [:error] [pid 11047] [client 127.0.0.1:45194]   * require_once("/home/marcus/workspace/cqu/cqu-mahara/htdocs/modul...") at /home/marcus/workspace/cqu/cqu-mahara/htdocs/module/objectfs/lib.php:503, referer: http://cqu-mahara-dev.marcus.local/admin/extensions/plugins.php
[Wed Jan 24 15:06:10.606102 2018] [:error] [pid 11047] [client 127.0.0.1:45194]   * PluginModuleObjectfs::get_client_components("client") at /home/marcus/workspace/cqu/cqu-mahara/htdocs/module/objectfs/lib.php:89, referer: http://cqu-mahara-dev.marcus.local/admin/extensions/plugins.php
[Wed Jan 24 15:06:10.606105 2018] [:error] [pid 11047] [client 127.0.0.1:45194]   * PluginModuleObjectfs::get_config_options() at /home/marcus/workspace/cqu/cqu-mahara/htdocs/lib/mahara.php:1779, referer: http://cqu-mahara-dev.marcus.local/admin/extensions/plugins.php
[Wed Jan 24 15:06:10.606108 2018] [:error] [pid 11047] [client 127.0.0.1:45194]   * call_static_method("PluginModuleObjectfs", "get_config_options") at /home/marcus/workspace/cqu/cqu-mahara/htdocs/admin/extensions/pluginconfig.php:52, referer: http://cqu-mahara-dev.marcus.local/admin/extensions/plugins.php
[Wed Jan 24 15:06:10.606112 2018] [:error] [pid 11047] [client 127.0.0.1:45194] , referer: http://cqu-mahara-dev.marcus.local/admin/extensions/plugins.php
[Wed Jan 24 15:06:10.606133 2018] [:error] [pid 11047] [client 127.0.0.1:45194] PHP Fatal error:  require_once(): Failed opening required '/home/marcus/workspace/cqu/cqu-mahara/htdocs/module/azure/vendor/autoload.php' (include_path='/home/marcus/workspace/cqu/cqu-mahara/htdocs/lib/:/home/marcus/workspace/cqu/cqu-mahara/htdocs/lib/pear/:.:/usr/share/php') in /home/marcus/workspace/cqu/cqu-mahara/htdocs/module/objectfs/classes/client/azure_client.php on line 15, referer: http://cqu-mahara-dev.marcus.local/admin/extensions/plugins.php